### ISSUE TITLE
Make use of :scheme and XForwardedProto clear and consistent in Envoy

### ISSUE THREAD

**Issue by PiotrSikora** [envoyproxy/envoy#14587]
It looks that `:scheme` pseudo-header is missing in HTTP/1.x requests, which makes it impossible to use route-based actions based on HTTP vs HTTPS scheme.

HTTP/1.1 request:
```
[C3][S17504616115359689975] request headers complete (end_stream=true):
':authority', 'localhost:10000'
':path', '/'
':method', 'GET'
'user-agent', 'curl/7.52.1'
'accept', '*/*'
```

HTTP/2 request:
```
[C2][S9968071387702895111] request headers complete (end_stream=true):
':method', 'GET'
':path', '/'
':scheme', 'http'
':authority', 'localhost:10000'
'user-agent', 'curl/7.52.1'
'accept', '*/*'
```

cc @howardjohn

**Comment by mattklein123** [comment]
Agreed we should be consistent here and fix this. cc @alyssawilk

**Comment by alyssawilk** [comment]
I thought the expectation was that ForwardedProto would represent the downstream scheme/security level internally, and :scheme was appended at the last minute for requests headed upstream, indicating the security of upstream requests.  Let's sort out a plan offline.

**Comment by PiotrSikora** [comment]
> I thought the expectation was that ForwardedProto would represent the downstream scheme/security level internally, and :scheme was appended at the last minute for requests headed upstream, indicating the security of upstream requests. Let's sort out a plan offline.

`:scheme` is already present and represents the downstream scheme in HTTP/2 requests, so we shouldn't have different semantics based on the downstream protocol version, especially when Envoy uses HTTP/2 pseudo-headers to represent other properties of HTTP/1.x requests (`:authority`, `:method` and `:path`).

**Comment by alyssawilk** [comment]
If filters use scheme, there's going to be issues where it means something different to upstream filters (where currently it would indicate upstream transport security) vs downstream filters (where it would be the downstream scheme if it exists, otherwise indicate downstream transport security).  This extra complexity is not present for :method or :path, which is why I would like to put some thought into what solution is least likely to cause confusion (cc @mattklein123 right back =P)

**Comment by mattklein123** [comment]
Yeah that's a good point that we swap `:scheme:` when we proxy upstream. Can we just clearly document that matching shouldn't be done on `:scheme` and it should be done on forwarded proto?

**Comment by alyssawilk** [comment]
That's what I've been telling people internally. 
If we think that's the proper thing to do (use x-forwarded-proto internally, serialize scheme right before encoding) I wonder if we should (very very carefully over several releases) remove scheme for http/2 to make it obvious to folks who are misusing it that they're using the wrong one?  Something like

* doc up that folks should use x-forwarded-proto
* add warnings when the scheme header is accessed
* make warnings fatal-by-default
* remove scheme header where we add x-forwarded-proto

?

**Comment by mattklein123** [comment]
I agree that ^ would be the right thing to do.

**Comment by alyssawilk** [comment]
I'll take on at least some of this when I get a chance.

**Comment by alyssawilk** [comment]
OK, having tried to standardize on ForwardedProto I am increasingly convinced that's going to cause problems.

So, new plan, go the other way.  

1) set :scheme for all requests.  For HTTP/1.1 it will be based on (trusted) forwarded proto, fully qualified urls, and encryption.  For HTTP/2 it'll be the scheme header.
2) Stop Envoy from sending :scheme header upstream based on upstream encryption which, to piotr's  point, seems to be changing the target URI. 
3) Envoy announce that folks should switch from x-forwarded-proto to :scheme as the source of truth for scheme.  Wait a month and/or a release
4) Stop setting X-Forwarded-Proto when not present, Remove X-Forwarded-Proto as a special header, only apply to HTTP/1.1 requests upstream.

All data plane changes would be flag guarded.  I think either 4) would have to go in its own release and/or we'd have to doc up how to safely upgrade with flags from (4) set false.

@snowp @antoniovicente @PiotrSikora SG?  I'd like explicit thumbs up from at least two of you before I dive into this.

**Comment by PiotrSikora** [comment]
SGTM. Thanks!

**Comment by mattklein123** [comment]
In general this sgtm but 2 things:

1) I know grpc has certain requirements with scheme. As long as scheme is set in all cases no matter what I think we are fine, but can we check this?
2) I'm very concerned that we will break people in the transition. Can we somehow warn if people are matching on XFP? Anything else we can do? I would also make the transition extra long with warnings first.

**Comment by alyssawilk** [comment]
yeah, I think this is going to be a multi-release fix, with lots of envoy-announce and warnings.  I plan on auditing all envoy use of XFP, and removing the XFP header accessors to help with downstream C++ filters but I don't think there's much we can do for wasm/lua other than envoy-annouce and slow guarded rollout.

**Comment by heatonmatthew** [comment]
We've just come across this change ([#15321](https://github.com/envoyproxy/envoy/pull/15321)) breaking our current implementation of gRPC upstreams.  Our downstream connection is encrypted, but our upstream is currently unencrypted.  

The upstream service is C# NET.Core 3.x which validates that the `:scheme` tag matches the actual incoming connection.  Since the `:scheme` header is now set to `https` to be consistent with the downstream but the actual incoming connection is not encrypted, the gRPC validation fails and rejects the request.

I'm not sure if this will be the same in other gRPC framework implementations but it seems problematic.

**Comment by alyssawilk** [comment]
Thanks for the heads up!

Thankfully you can obviously flip the runtime flag back to address your problem, while you sort out what workaround makes more sense, but I don't think this warrants delay of deprecating this runtime guard.  If neither of the suggestions below works for you, please do loop back through.

Fundamentally, Envoy forwarding the scheme of the original URI is the correct thing to do.  if  C# NET.Core 3.x requires https requests be sent over TLS you can either add host rewrites to downgrade the scheme to http, or do TLS to upstream.  I know it's tempting to keep traffic unencrypted on your own local networks, but I can tell you from personal experience that's not always safe to do even if you own the wires being transited (https://blog.encrypt.me/2013/11/05/ssl-added-and-removed-here-nsa-smiley/) so I strongly encourage doing the latter.

**Comment by heatonmatthew** [comment]
Thanks alyssiawilk.  I understand that this is a complicated issue with lots of different considerations. There are some aspects here that I have some questions about.

Firstly, it seems from reading the conversation that this change was implemented to enable the internal routing configuration within Envoy to consistently detect that the downstream connection was encrypted/unencrypted.  Primarily this started with the deficiency that the `:scheme` pseudo-header was not simulated on HTTP/1.1 connections.  It now is simulated which enables routing configuration to internally use `:scheme` regardless of the incoming version of HTTP.

Secondly, this change has now meant that the upstream `:scheme` header is set to the downstream one (even if it has been simulated/inferred for HTTP/1.1 connections).  To your point on the 29 Jan that changing the `:scheme` fundamentally changes the target URI, that is true.  So internally not doing this makes routing a challenge.  However, downgrading the connection from TLS to unencrypted when proxying upstream does actually change the target URI by definition at the outgoing side.  Therefore it _may?_ make sense to do a corresponding downgrade of the `:scheme` header at the egress point (i.e. after the internal filters are processed) so that it is consistent with the downgraded connection being made upstream.

Regardless, our plan is to encrypt internal comms.  We had planned to implement a Service Mesh (with Kubernetes sidecars) to transparently handle the encryption of traffic for workloads (including those of 3rd software components for which we _might_ not have control over their TLS capabilities).  However, if we implement a Service Mesh to perform the encryption for us, the traffic is effectively tunneled and the target workload receives the protocol scheme initiated from the egress point of Envoy before it goes into the service mesh (i.e. the service mesh sidecar removes the encrypted session before proxying the original connection scheme to the workload).  If our only option to have consistency between the `:scheme` tag and the actual connection arriving at the workload is to use a TLS connection, then we effectively will need to be running a TLS gRPC connection over a second encryption layer provided by the service mesh.  This seems a little redundant.

We will find a solution either way, but thought it might be important to consider the implication for other Envoy customers.

**Comment by mattklein123** [comment]
@alyssawilk unfortunately ^ is what I thought might happen IMO we can't break people with this change even if what they are doing is "wrong." Should we have this behavior be configurable?

**Comment by alyssawilk** [comment]
so @mattklein123 I agree we need to support incorrect behavior, but I don't think we need another knob.  
If folks are sending cleartext requests and want https scheme, I think using the existing scheme_redirect to change their actual http to https is the workaround I'd advise.

Given we have that knob already, I was hoping @heatonmatthew would check it out, and then if it works it can be the canonical workaround for folks who encounter any problems.

**Comment by heatonmatthew** [comment]
Thanks all.  @alyssawilk, unfortunately, I'm using Envoy as an embedded component within the Gloo Edge product. I will liaise with them on how I might set the flag to test this out.

However, after thinking on this over the weekend, the Service Mesh use case still seems problematic and likely to cause additional issues to be logged.  It might be worth seeking some input from the Istio project about whether this change will cause issues for their standard [sidecar Security Architecture](https://github.com/envoyproxy/envoy/issues/14587#issuecomment-828832216) where the Envoy proxies perform the secure Identity, Authentication & Authorisation functions between Pods while enabling the internal comms between containers in the Pod to have lower requirements.

**Comment by alyssawilk** [comment]
cc @lambdai @lizan @tbarrella

**Comment by heatonmatthew** [comment]
I've confirmed that using the `envoy.reloadable_features.preserve_downstream_scheme` resolves the issue I've encountered.

**Comment by heatonmatthew** [comment]
I had a quick play yesterday with route-based scheme rewrites and didn't succeed but I will put some more effort into it and get back to you.

I've been reading through the attached issues and have some questions that I'd like to ask.  Is this the appropriate forum for me to do so?  They are regarding:

- The interpretation of clause 8.1.2.3 of [rfc7540](https://tools.ietf.org/html/rfc7540#section-8.1.2) that was taken in #14816 and security assumptions implied that might not be true.
- A specific service-mesh scenario that uses a downgrade of HTTPS to HTTP for some Kubernetes workloads but for which none of the unencrypted traffic actually hits the network.  I have a feeling that @mattklein123 was asking questions similar to this in one of the reviews but the back and forth is hard to follow.  I believe this scenario/implementation improves security posture by enabling the Service Mesh to handle certificate management activities on behalf of backend microservice workloads without needing the implementers to manage a certificate distribution/rotation/update infrastructure themselves (and potentially introducing attack vectors into the certificate management infrastructure by not doing that correctly/well).
- If the service-mesh scenario above is valid, then this change seems to impose a double-configuration inefficiency/complication such that the encryption level has to be specified on the last-hop connection and also scheme header re-writes configured.

I'm not keen on getting into a big argument but can provide more details if it is appropriate.  Thanks.

**Comment by alyssawilk** [comment]
No argument needed! I should have been more clear that the 6 month thing is more of a default bound on how long we have to find that workaround and make sure it's solid - it is extensible but I'm sure we'll find a way to make sure your deployment works without flipping that guard back.   I'm not clear today what that will look like, to both fix your use case and ensure we don't have a giant security hole for the edge proxy case but we'll find something :-)

I'm out today (just checking in to land the H3 alpha) but if you're around next week I'd be happy to sync in real time, to get a better understanding of your deployment set-up and either help sort out the rewrite case or find something else that works.  Any chance you're on envoy slack?  If so I can ping over there on Monday

**Comment by heatonmatthew** [comment]
Looking around more broadly I can see that this support will be needed for a long time.  The .NET team agree that the spec is ambiguous but are working on the assumption that all proxies will need to support the previous behaviour as a configuration option into the future (see [comment](https://github.com/dotnet/aspnetcore/issues/30532#issuecomment-855016549)).

I also note that the following workarounds are either in place or being requested:

- Google - https://issuetracker.google.com/issues/187470251?pli=1
- AWS NLB - https://github.com/dotnet/aspnetcore/issues/30532
- Istio - https://github.com/istio/istio/issues/33136

If support was provided for a specific scheme rewrite option for each route, I believe that it will enable the desired outcome.  What does concern me is that each solution (e.g. Google, AWS NLB, Istio, Gloo etc.) will need to either provide a mechanism to detect the downgrade and set the scheme rewrite for each route as part of their xDS plane implementation, or pass the additional configuration required onto their end-users.  

Deliberating on how they want to support the change is likely to take those teams a while, so I suspect that pushing out the timeframe for removal of the `envoy.reloadable_features.preserve_downstream_scheme` flag makes sense to me.

**Comment by mattklein123** [comment]
> Deliberating on how they want to support the change is likely to take those teams a while, so I suspect that pushing out the timeframe for removal of the envoy.reloadable_features.preserve_downstream_scheme flag makes sense to me.

I would like to hear from @alyssawilk on this, but I continue to be concerned that we may be causing undo hardship on our users with this change, and I'm wondering if there are any compromise that are possible. (I don't know what that would be as I haven't spent enough time fully understanding this issue, and @alyssawilk has spent a ton of time, so hopefully he can summarize and educate us on the current thinking.)

**Comment by qiwzhang** [comment]
Here I like to summarize the issue ESPv2 is facing with this change:

Setting correct :scheme header is important for Google IAP to work. Google IAP requires :scheme to be https, otherwise it will send 302 redirect.
When ESPv2 envoy is deployed in Google Cloud-Run service, its downstream protocol is http since it is getting request from GFE and GFE terminates the SSL from the client, and its upstream uses https, they could be any of Google serverless platforms, e.g. app-engine, Cloud-function, or Cloud-Run. "preserve_downstream_scheme" behaviors is not correct for ESPv2.
ESPv2 will like to revert back to the old behaviors that setting :scheme based on the upstream security level.

So I propose to add a config flag somewhere to keep the old behaviors.

**Comment by alyssawilk** [comment]
So Matt and I I discussed this offline and think the way to go about this is have a per-route scheme modifier, as we do for host and path, and also have a per-HCM option to set scheme.  I think for ESPv2 you'd just set that to https, and get the behavior you want.   Does that sound reasonable?

**Comment by qiwzhang** [comment]
Sound Good.  Thanks

**Comment by alyssawilk** [comment]
Ok, way back in January I declared the last step as not adding XFP, as a forcing function to get folks to migrate over to :scheme
I think that's going to be more trouble than it's worth (it's _definitely_ going to break a ton of filters and while they're probably occasionally doing the wrong thing breaking them isn't great) so I think I'm going to stop at :scheme being always set and the new docs informing which should be used when, and letting folks have their buggy filters (which now that we're sending :scheme correctly are only buggy for the .0001% of traffic that XFP and :scheme don't match)

**Comment by johnlanni** [comment]
I'm still a little confused about this mechanism.
Based on the above discussion, the XFP header will retain settings in the future, so this header itself can be used to indicate whether envoy's downstream has TLS enabled. There is no need to use the downstream scheme header as the upstream scheme header, if this scheme header is transparently transmitted , causing grpc frameworks such as .NET to be unable to handle it normally.
Although HCM provides options to override the scheme, this is not enough, because what we actually need is to use the upstream secure state to set the scheme header like the previous version do:

https://github.com/envoyproxy/envoy/blob/345ffe37148b7a35b6e8e04db0300463689e3ff1/source/common/router/router.cc#L93-L102

**Comment by johnlanni** [comment]
Hi @alyssawilk , thanks for you reply. My scenario is that Envoy receives gRPCs(encrypted) requests and proxies gRPC (not encrypted) request to  a .NET service. The .NET gRPC framework checks the `:scheme` header, if the request is not encrypted, but the `:scheme` is `https`, the request will fail.

I solved the problem by setting the HCM option `scheme_header_transformation` to `http`, but this option has a side effect of setting the XFP header to `http`, which makes it impossible to determine the Envoy downstream protocol through the XFP header.

**Comment by johnlanni** [comment]
> the mechanism boils down to
> 
> 1. if we trust the source, we use XFP
> 2. if we don't, we set XFP based on the scheme of the conneciton
> 
> This seems to have worked for all users for a couple of years now. If you have a concrete example of it not working let us know and we'll try to address.

This mechanism is reasonable, but I think the `:scheme` header of the upstream request should not be set to the same as XFP, because this makes `:scheme` ambiguous, is the current request‘s scheme or the original request's scheme? I prefer to define it this way:
`:scheme`: The currently request's scheme.
`x-forwarded-proto`: The original request's scheme.

**Comment by alyssawilk** [comment]
yeah unfortunately many servers predate the x-forwarded-proto header, and so expect scheme to be the scheme of the original request.  I think always setting scheme based on upstream crypto instead of defaulting to XFP is a reasonable request for a cluster configuration option but you'd probably need to find someone to sign on to implement it.  Worth filing as its own feature request if you're interested.
